File: smells/unclear/indecisive.js
/* Indecisive Tests
 *
 * primitive obsession / lack of good test controls
 */

// Тестируемый модуль
function joinPath (fragments) {
  var separator, pattern
  if (process.platform === 'win32') {
    separator = '\\'
    pattern = /\\+/g
  } else {
    separator = '/'
    pattern = /\/+/g
  }
  return fragments.join(separator).replace(pattern, separator)
}

// Тесты
module.exports = {
  simpleCase: function () {
    var fragments = ['foo', 'bar', 'baz']

    var result = joinPath(fragments)

    if (process.platform === 'win32') {
      assert.equal(result, 'foo\\bar\\baz')
    } else {
      assert.equal(result, 'foo/bar/baz')
    }
  },
  containsSeparators: function () {
    var fragments
    if (process.platform === 'win32') {
      fragments = ['\\foo\\', 'bar\\biz', 'baz\\']
    } else {
      fragments = ['/foo/', 'bar/biz', 'baz/']
    }

    var result = joinPath(fragments)

    if (process.platform === 'win32') {
      assert.equal(result, '\\foo\\bar\\biz\\baz\\')
    } else {
      assert.equal(result, '/foo/bar/biz/baz/')
    }
  }
}

File: smells/unclear/indecisive/indecisive.rb
# Subject under test
require 'rbconfig'

def join_path(fragments)
  if /mswin/ =~ RbConfig::CONFIG['host_os']
    separator = "\\"
    pattern = /\\+/
  else
    separator = "/"
    pattern = /\/+/
  end
  fragments.join(separator).gsub(pattern, separator)
end

# Test
class Indecisive < SmellTest
  def test_simple_case
    fragments = ["foo", "bar", "baz"]

    result = join_path(fragments)

    if /mswin/ =~ RbConfig::CONFIG['host_os']
      assert_equal "foo\\bar\\baz", result
    else
      assert_equal "foo/bar/baz", result
    end
  end

  def test_contains_separators
    if /mswin/ =~ RbConfig::CONFIG['host_os']
      fragments = ["\\foo\\", "bar\\biz", "baz\\"]
    else
      fragments = ["/foo/", "bar/biz", "baz/"]
    end

    result = join_path(fragments)

    if /mswin/ =~ RbConfig::CONFIG['host_os']
      assert_equal "\\foo\\bar\\biz\\baz\\", result
    else
      assert_equal "/foo/bar/biz/baz/", result
    end
  end
end
